### ISSUE TITLE
Issue reporting form needs refinement Version 1

### ISSUE THREAD

**Issue by josephj11** [Elliria/autokey-fork#4]
### Has this issue already been reported?

- [X] I have searched through the existing issues.

### What type of issue is this?

Enhancement

### Which Linux distribution did you use?

N/A

### Which AutoKey version did you use?

N/A

### Can you briefly describe the issue?

Suggested changes/improvements for this form

### What are the steps to reproduce the issue?

Select New issue

### Anything else?

YAML Issue Form Notes:

The following are suggestions of what I'd like to see. I do not know if they are possible or how difficult they might be to imlement.

Bug Report could say Open a new issue using our issue reporting form

Don’t see your issue here? Open a blank issue.
is the first thing I clicked on. I wouldn't be here if I thought the issue already existed and there are no issues on this page to "see".

Could say: Skip our Issue reporting form and open a generic issue

Issue types:
Documentation should be a valid issue type (new)
Since a new feature is a type of enhancement, maybe it could be eliminated - just thinking.
Default might be better as Unspecified rather than None
UI/UX instead of UI/Usability ?

Headings are fine. Supporting text font is too small.

Can AutoKey Version be prefilled with AutoKey 0.9?
Input box is huge, resize?

How did you install:
Default might be better as Unspecified rather than None
Add N/A - not applicable

Package Manager is not a good choice because it's not specific enough. It needs to open a pop-up for input as to which package manager was used or direct the user to fill in the details somewhere else.
This is a problem because:
Debian - using dpkg works, but doesn't address any dependencies
Arch - has at least two package management systems (aur, pacmd) that don't do the same things

PPA - same as package manager - anybody can create a PPA, so we need to know which one was used. We don't even have one of our own now.

Can the issue be reproduced?
Default might be better as Unspecified rather than None
Add N/A - not applicable

Can you provide the output of the AutoKey command?
Needs refinement:
We only need this for some bugs.
Output from starting AutoKey from a terminal
Sometimes the messages are sufficient without a full trace.

If your issue involves a crash/throwing an exception, please recreate your issue by running AutoKey from a terminal window and paste any messages it issues below.

If AutoKey produced unexpected results, often running an AutoKey trace and pasting the results below will be helpful.

More [details](https://github.com/autokey/autokey-python2/wiki/Problem-Reporting-Guide) on how to do this.

**Comment by josephj11** [comment]
Just noticed that it says I added the bug and help wanted labels to this issue. I did not do this manually and both are incorrect for this particular issue. Maybe automatically assigning a label of something like "new issue" would be appropriate - indicating that no one has looked at it yet and it needs attention. (This would be manually deselected after triage...)  

Similarly, it says I assigned the issue to Elliria which I also did not manually do and wouldn't presume to do without discussing it with her first...

**Comment by Elliria** [comment]
## Done:
These issues are done, but feel free to comment on any of them or make further suggestions, especially on that last one which will probably need more work.
1. Replace the "Bug Report" title with the "Open a new issue using our issue-reporting form" title.
2. Remove Elliria as the default assignee for new bugs.
3. Replace default "bug" and "help wanted" labels with "new issue" label.
4. "What type of issue is this?" section: Add the "Documentation" option to the dropdown.
5. "What type of issue is this?" section: Remove the "Feature (new)" option from the dropdown since it's a type of enhancement.
6. "How did you install?" section: Add N/A - not applicable
7. "Can the issue be reproduced?" section: Add N/A - not applicable
8. "Can you provide the output of the AutoKey command?" section: Needs refinement.

**Comment by Elliria** [comment]
## Out of my control - controlled by GitHub:
I don't think we have the power to do anything about these, but I'll continue to do research to see if there are ways around it. Also, we could file bug reports with GitHub to see if they'll hand us a bit more power.
1. Rename "Don’t see your issue here? Open a blank issue." at top of form with "Skip our Issue reporting form and open a generic issue" or maybe get rid of it altogether.
2. Headings are fine. Supporting text font is too small.
3. Input box is huge. Resize?

**Comment by Elliria** [comment]
## Out of my control - controlled by GitHub or YAML:
I'm not sure if it's GitHub or YAML that's the roadblock to these, but in any case, I don't think we hve the power to do anything about these since they're being done for us and there are no available valid YAML keys that will get the job done. This is another situation in which filing a bug report with GitHub might get us somewhere.
1. Issue types: Use the "Unspecified" option as the default for the dropdown instead of the "None" option.
2. "How did you install?" section: Use the "Unspecified" option as the default for the dropdown instead of the "None" option.
3. "Can the issue be reproduced?" section: Use the "Unspecified" option as the default for the dropdown instead of the "None" option.

**Comment by Elliria** [comment]
## Up for discussion:
1. Issue types: Possibly use the "UI/UX" option instead of the "UI/Usability" option in the dropdown?

	_How about "Interface" or "User interface" instead? We already have a "User Interface" label, but it was never mentioned or used as a classification type in the original AutoKey bug report form and it's been used in the past._

2. Can AutoKey Version be prefilled with AutoKey 0.9?

	_Yes. It can either be inserted for you as text in the textarea or there can be a placeholder in the textarea (like the one in the steps to reproduce section). If it's inserted text, you can delete and replace it, if needed. If it's a placeholder, it disappears when you start typing, but acts as a prompt until you do. I'm hesitant to suggest a specific version number, though, because if there's one there, folks might be tempted to simply accept it rather than checking._

3. "How did you install?" section: The "package manager" option isn't specific enough and we need a pop-up for input.

	_Would a textarea be better than a dropdown for this?_

**Comment by josephj11** [comment]
1, 2, 3) See if the links on Gitter help with specifying default values for some fields.

I found a bunch of font stuff, but it only seems to apply to something called R Markdown. There was something else about referencing a custom CSS near the start of the YAML. Both are things I don't have any experience with.

1a) I think of the User Interface label as primarily referring to the UI code, but I have never discussed this with anyone else. UI/UX feels more generic like look and feel rather than the actual nuts and bolts, but that's just me.

2a) Valid point. Might work as one of those prompts that disappears when you type in the field 

3a) Package manager as a text box would probably be better.

**Comment by Elliria** [comment]
I saw some R Markdown stuff, too, but I don't think GitHub uses it. I should say I don't think GitHub lets us use it. I did a full search of this entire document and wasn't able to find any mention of R Markdown: [https://docs.github.com/en/github/writing-on-github](https://docs.github.com/en/github/writing-on-github), but it's still worth looking into.

I brought up the UI/UX label in Gitter chat, so maybe we'll get some input.

In looking over the stuff above, I realized that I shouldn't have said "package manager" above, but instead something like "distribution repository" or whatever the simplest way of referring to each variant of GNU/Linux' default package store (in other words, whichever version of AutoKey comes with your distribution). Anyway, in the event that that's not something we would want to use as an option, would replacing the "How did you install AutoKey?" dropdown entirely with a textarea that contains a placeholder as a prompt be okay? I don't think a textarea can be used as an option, at least on GitHub.

**Comment by josephj11** [comment]
Good catch. "distribution repository" comes much closer to correct as the package managers are just tool for accessing them.

Not totally clear on what the issue is, but replacing the dropdown with a text area with a detailed placeholder sounds better than what we have now.

Tried searching Gitter for ux (text:ux) - didn't find your post about UI/UX.

 Found this!: https://github.community

Only spent a few minutes there, but putting yaml into the search field produced a large number of hits.

https://github.community/t/getting-started-with-the-community-forum/186591

**Comment by Elliria** [comment]
Yeah, I should have known better than to use "package manager" there. Ah, well, it's fixed now.

The dropdown has been replaced. I also changed "Documentation Issue" to "Documentation" in the classification dropdown. The bug report form should be shaping up pretty nicely within the abilities given to us by GitHub, and unless you or anyone else has additional suggestions for what ought to be done to it, the only thing holding it back now is whether there will be any further changes to the AutoKey labels.

The UI/UX mention was in the message with the huge **"Bug report label review"** heading if you scroll up a bit in the Gitter chat from the bottom. It used the actual wording in the labels rather than UI/UX, which was why that search didn't turn it up. In that message, I was most interested in the differences between what's in the current bug reports and what's in the project's labels and was hoping to synchronize them as much as possible without doing harm to any that are and/or have been in use.

Thanks for the link to the community. It's a shame that you're not automatically logged in there when you're logged in here. I'm never comfortable with the account linking that it seems to want, so I'm not going to post there. I'll ask my YAML question about default dropdown entries on Stack Overflow or Stack Exchange. Meanwhile, I'll climb around in those YAML hits just to see if there's something I can learn that might be useful for AutoKey.

**Comment by josephj11** [comment]
Great progress. More notes:

The initial message (before they select New Issue) talks exclusively about bugs. Should probably say issues because it's for enhancements too.

I'd really like it if this part would also say: this isn't for support questions. For those, go to our forum or Gitter.
I have prettier wording for that - see below the line. (If you copy it, get it in source form so you don't lose the hyperlinks.)

In the form:
Is Title optional? I don't see a red x ...

Rest looks very good. We just have to figure out how to get some feedback from the folks on Gitter.

----

BTW: Whenever you have a question (as opposed to an issue) it's better to post it on our [list](https://groups.google.com/forum/#!forum/autokey-users) or on [Gitter](https://gitter.im/autokey/autokey). The list is more user oriented whereas more technical questions will get a better answer on Gitter.

Both of these venues have a number of active users, so you are likely to get more, better, and faster answers on them than here on GitHub where only our very busy developers and a few users see the posts. We also prefer to have the developers developing rather than taking time to answer questions and manage issues that the user community could handle for them.

**Comment by Elliria** [comment]
I reworded the initial message (before users select New Issue) to remove "bug" from it and also added "Please" to it.

I can't add your text to it becuase GitHub limits us to 200 characters for that description field (it's currently at 188 characters). What I did instead is to add another required checkbox that used your suggested text to point out that there are better venues for questions than these issue reports and asks the user to confirm that this is not a question. I tucked it in just before the **"What type of issue is this?"** section. Hopefully that will work just as well. We're not limited in size on those, so if you'd like any edits to be made to that section, we can do that.

The title is not optional and I don't see a red x either, but that's controlled by GitHub. We should file an issue report with them on that. If nothing else, it should have a red border until it's filled out.

Meanwhile, if you like, we can populate that field for the user with a default issue report title that they can change. The reason I didn't do that is that folks might just leave it and move on and not even notice that it needs to be edited, so you'd suddenly end up having to retitle issue reports. Titles are easy to edit, though, so something like that could solve it neatly.

Note that we can't put any introductory or instructional text before the title box. All we have is the form's 200-character description field we were talking about above and everything beneath the title. We've got access to the title field in the code, but can only specify its value as a string. Placeholders aren't accepted. I gave the title field the value of ```[New issue]``` so you can see how that behaves when filling out an issue report. We can change that to whatever you like.

Any suggestions you have for getting feedback on the labels is welcome. I've tried a couple of times in Gitter. It should be up to the folks who use the labels. If we get no feedback, we could do our best with what we're scheming up here and always make changes to them later if anyone pipes up. I'll do some tests today to see what happens to links if we change labels.

**Comment by josephj11** [comment]
This is working out far better than I was hoping for. Great work!

If the Title is required, it should probably be left empty. I was going to suggest a placeholder, but you already tried that. I definitely don't want to be putting words in the users mouths by changing "new issue" to whatever I think it should be. I usually request that they change the title if I think it could be better although I have edited a few of them.

If we don't get more feedback, we just issue a "don't complain later" statement and install it when we're ready.
A universal law is: Users usually don't know what they want, but they do know what they don't want. LOL
I'll review the form as soon as I get a chance.

**Comment by josephj11** [comment]
Just reviewed the form. Looks great to me except for the pre-filled title.

**Comment by Elliria** [comment]
Okay, the title key/value pair is gone now. I like your idea of the "don't complain later" part, or at maybe we can just expect potential complaints later and will welcome them as feedback.

**Comment by Elliria** [comment]
By the way, I just now did a test by adding the original Markdown version of the issue template back into this repository. I wanted to know if it needed to be deleted (the way I had done in here) or if it could remain in place while allowing the YAML version to be used by default. It looks like if there are both versions, the YAML one will be preferred, so the old template won't even need to be deleted.

**Comment by josephj11** [comment]
The "don't complain later" thing is mostly for "Why did you do that? ..." LOL. Of course we'll accept feedback whenever we get it as long as it's constructive.

**Comment by Elliria** [comment]
Sounds good. Should I do a pull request at this point or is there anything else that needs to be done first?

**Comment by josephj11** [comment]
I think a PR would be great. When you make one, it will have to be reviewed and merged by @BlueDrink99 (or by me in  the last resort), so that should get it some attention (although BD has already looked in on this). Once you do the PR, you can post on Gitter and on the forum with something like "This is the way it's going to be unless anyone wants to suggest some improvements before we merge it."

**Comment by Elliria** [comment]
I might not be able to quite do that. I think I'll start a bit more gently by posting that it's ready for a PR and offering to let everyone come take a look before I go ahead with the PR.

**Comment by josephj11** [comment]
However you want to proceed is fine with me. Your approach is more diplomatic. I was (over?) responding to the lack of feedback so far.
